### PR TITLE
[Feature] Error mitigation structure + ZNE for Pulser backend

### LINKED ISSUE #48: Simulate readout errors
This allows to emulate the presence of readout errors in a shot-based simulation.

### PR DESCRIPTION
**Pull request by Roland-djee**
Closes #48 and #123 and #121

- [x] ZNE for analog blocks, Pulser backend only.
- [x] Readout mitigation.

### CODE REVIEW

**Comment by Roland-djee** [comment]
@gvelikova Feel free to build on top of this.

**Comment by madagra** [comment]
@gvelikova @Roland-djee let's add here only the simulated readout error and then go for the error mitigation side in a separate MR.

**Comment by Roland-djee** [comment]
> @gvelikova @Roland-djee let's add here only the simulated readout error and then go for the error mitigation side in a separate MR.

I guess what you suggested is the purpose of #128.  Probably this PR should be renamed.

**Comment by Roland-djee** [comment]
Hey @madagra I think the ZNE part is ready for you to review. Will work on the readout mitigation protocol next.

**Review by gvelikova — APPROVED** [review]
Looks good to me. The protocol -> measurement (e.g., in QuantumModel) in the error models PR, but this shouldn't break things after both PRs are merged. I guess some test cases have to be added.

**Comment by Roland-djee** [comment]
> Looks good to me. The protocol -> measurement (e.g., in QuantumModel) in the error models PR, but this shouldn't break things after both PRs are merged. I guess some test cases have to be added.

Thanks. Yes, I'd rather merge this one when yours is merged to fix potential issues here. In the same time, I'll add few more test cases indeed.

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:59]: I honestly don't understand why we have this. The noise model should be something that the user supplies externally via the backend configuration and not in the mitigation routine. Here you can inspect the `sim_config` object of the input backend to check the noise model but otherwise, it does not really make sense to add the noise directly here. What if you want to extrapolate for real QPU executions, for example?

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:63]: I suggested to remove this `run_noisy`, now I see how you use it. To make it cleaner, I would define a `run_dm` function in the backend which returns a density matrix instead of the wavefunction. Only the Pulser backend would implement it, thus making this function fail if the Pulser backend is not used + it makes it also a bit more flexible for future updates (e.g. at some point also PyQ will implement a `run_dm` method).

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:65]: Remove.

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:76]: Are batched dimensions supported as well, right?

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:77]: The `analog_zne` looks like a very basic implementation of the zero noise extrapolation scheme. Isn't there any unfolding happening? I see only the final fit (extrapolation) to zero noise.

File: qadence/mitigations/zne.py
Comment (madagra) [inline qadence/mitigations/zne.py:89]: If you properly implement the `run_noisy` --> `run_dm` method this should not be necessary.

**Review by madagra — COMMENTED** [review]
In general, it looks like the right direction to me. However, I have several design comments and a more general algorithmic one.

**Comment by madagra** [comment]
@Roland-djee for the readout error mitigation, do you already have the routine somewhere?

File: qadence/backend.py
Comment (gvelikova) [reply qadence/backend.py:25]: Or at least Mitigation? ErrorMitigation sounds the most informative.

Comment (Roland-djee) [reply qadence/backend.py:25]: I stuck to `Mitigations` for now but we can rename it later if we redefine the interface a little.

File: qadence/backends/pulser/backend.py
Comment (Roland-djee) [reply qadence/backends/pulser/backend.py]: There is now both the `noise` and `mitigations` modules.

File: qadence/mitigations/protocols.py
Comment (Roland-djee) [reply qadence/mitigations/protocols.py:19]: #183

File: qadence/noise/readout.py
Comment (gvelikova) [inline qadence/noise/readout.py:1]: * or Poisson.

File: qadence/types.py
Comment (Roland-djee) [reply qadence/types.py:61]: Indeed.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:44]: Simplified.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:65]: Done.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:77]: #181

**Comment by Roland-djee** [comment]
@madagra I'll address your comments about the ZNE tomorrow. If you've got some spare time, I'd be happy to hear your thoughts about the readout mitigation. Thanks.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:61]: Instead of adding yet another method to the backends, would it make sense to add an argument to query for the output of run ? Say, `Output.WF` or `Output.DENSITY` ? That was also a query from yesterday's meeting. @madagra ?

Comment (Roland-djee) [reply qadence/mitigations/zne.py:61]: Or a `Result` type would be even cleaner I think.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:1]: Done.

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:85]: In #187

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:61]: In #187

File: qadence/mitigations/zne.py
Comment (Roland-djee) [reply qadence/mitigations/zne.py:72]: In #187

**Comment by Roland-djee** [comment]
Hey @madagra and @gvelikova if you're fine with this PR, I'll merge it as soon as the pipeline passes and work on #187 where all the remaining comments will be addressed.

**Pull request merged**
